src/main/java/jenkins/plugins/openstack/compute/SlaveOptionsDescriptor.java

```diff
@@ -242,49 +249,153 @@ public FormValidation doCheckHardwareId(
-
+        final String valueOrEmpty = Util.fixNull(imageId);
+        boolean existingValueFound = valueOrEmpty.isEmpty();
+        m.add(new ListBoxModel.Option("None specified", "", existingValueFound));
         try {
-            final Openstack openstack = Openstack.Factory.get(endPointUrl, identity, credential, zone);
-            for (Image image : openstack.getSortedImages()) {
-                String name = image.getName();
-                  if (Util.fixEmpty(name) == null) {
-                    name = image.getId();
+            final JCloudsCloud.BootSource effectiveBS = calcBootSource(bootSource, defBootSource);
+            if (effectiveBS!=null && haveServerDetails(endPointUrl, identity, credential, zone)) {
+                final Openstack openstack = Openstack.Factory.get(endPointUrl, identity, credential, zone);
+                switch(effectiveBS) {
+                    case IMAGE:
+                        final Map<String, Collection<Image>> images = openstack.getImages();
+                        for (String value : images.keySet()) {
+                            final String displayText = effectiveBS.toDisplayName() + " " + value;
+                            m.add(displayText, value);
+                            existingValueFound |= valueOrEmpty.equals(value);
+                        }
+                        break;
+                    case VOLUMESNAPSHOT:
+                        final Map<String, Collection<VolumeSnapshot>> volumeSnapshots = openstack.getVolumeSnapshots();
+                        for (String value : volumeSnapshots.keySet()) {
+                            final String displayText = effectiveBS.toDisplayName() + " " + value;
+                            m.add(displayText, value);
+                            existingValueFound |= valueOrEmpty.equals(value);
+                        }
+                        break;
+                    default:
+                        break;
                 }
-                m.add(name);
             }
-            return m;
         } catch (AuthenticationException | FormValidation | ConnectionException ex) {
             LOGGER.log(Level.FINEST, "Openstack call failed", ex);
         } catch (Exception ex) {
             LOGGER.log(Level.SEVERE, ex.getMessage(), ex);
         }
-
-        if (Util.fixEmpty(imageId) != null) {
+        if (!existingValueFound) {
             m.add(imageId);
         }
-
         return m;
     }
 
     @Restricted(DoNotUse.class)
     public FormValidation doCheckImageId(
             @QueryParameter String value,
-            @RelativePath("../../slaveOptions") @QueryParameter("imageId") String def
+            @RelativePath("../../slaveOptions") @QueryParameter("imageId") String def,
+            @QueryParameter String bootSource,
+            @RelativePath("../../slaveOptions") @QueryParameter("copyOfBootSource") String defBootSource,
+            // authentication fields can be in two places relative to us.
```

Comment:
`@InjectOsAuth` should do it. I am sorry you have to go through the pains to discover how messed up this all is.